src/utils/encoding.ts

```diff
@@ -0,0 +1,56 @@
+import fs from "fs";
+import iconv from "iconv-lite";
+import readline from "readline";
+import { extractFileFormat } from "./file";
+import type { SupportedImportEncoding } from "./file";
+
+export const isMismatchEncoding = async (
+  filePath: string,
+  encoding: SupportedImportEncoding,
+): Promise<boolean> => {
+  const format = extractFileFormat(filePath);
+  let isMismatch = false;
+  switch (format) {
+    case "csv":
+      isMismatch = await isMismatchEncodingOfCsvFile(filePath, encoding);
+      break;
+  }
+
+  return isMismatch;
+};
+
+const isMismatchEncodingOfCsvFile: (
+  filePath: string,
+  encoding: SupportedImportEncoding,
+) => Promise<boolean> = async (filePath, encoding) => {
+  const decodedFirstLine = await getDecodedFirstLine(filePath, encoding);
+  return isContainUntranslatableChars(decodedFirstLine);
+};
+
+const getDecodedFirstLine: (
+  filePath: string,
+  encoding: SupportedImportEncoding,
+) => Promise<string> = async (filePath, encoding) => {
+  const stream = fs.createReadStream(filePath);
+  const readWriteStream = stream.pipe(iconv.decodeStream(encoding));
+  const reader = readline.createInterface({
+    input: readWriteStream,
+  });
+
+  return new Promise((resolve, reject) => {
+    reader.on("line", (line: string) => {
+      reader.close();
+      resolve(line);
+    });
+
+    reader.on("error", reject);
+    stream.on("error", reject);
+  });
```

Comment:
Because the linebreak character is common in any encoding, we can reorder the process to simplify this function.
(Also, preventing the error described in the comment for line 35. )

1. Read one line from the input file
  - We can use `filehandle.readLines()` and Async Iterator to simplify the code
  - readLines: https://nodejs.org/api/fs.html#filehandlereadlinesoptions
  - The return value of readLines has `Symbol.asyncIterator()` so we can use `for await...of` or `next()` to read lines
2. Decode lines
  - Because we already pick only one line, we can use the synchronous decode function

Reply:
@mshrtsr 

Thank you for your suggestion.

I see `filehandle.readLines()` is a convenient method to create a `readline` interface. Same as the package I use https://nodejs.org/api/readline.html#class-interfaceconstructor.

So, they're the same. Can you check it again?
However, I refactored to use `Symbol.asyncIterator()`.

In addition, I tried to use `iconv.decode` (without `stream.pipe(iconv.decodeStream(encoding));`), but the decoded string is incorrect.
Now, I don't know the reason yet. I will check it more.

Reply:
@hung-cybo 
Yes, there is no semantic difference.
I prefer to use a convenient method if possible, but both are OK.

In conclusion, we decided to keep using `iconv.decodeStream()`.
So we cannot use `filehandle.readLines()` in this code.
https://github.com/kintone/cli-kintone/pull/437#discussion_r1301196095

So we can keep this code as is 👍